pkg/streamer/reader.go

```diff
@@ -605,29 +605,58 @@ func (r *BinlogReader) parseFile(
 		return false, false, latestPos, "", "", nil
 	}
 
-	needSwitch, needReParse, nextUUID, nextBinlogName, err = needSwitchSubDir(r.cfg.RelayDir, currentUUID, fullPath, latestPos, r.uuids)
-	if err != nil {
-		return false, false, 0, "", "", err
-	} else if needReParse {
-		// need to re-parse the current relay log file
-		return false, true, latestPos, "", "", nil
-	} else if needSwitch {
-		// need to switch to next relay sub directory
-		return true, false, 0, nextUUID, nextBinlogName, nil
-	}
+	switchCh := make(chan SwitchPath, 1)
+	switchErrCh := make(chan error, 1)
+	updatePathCh := make(chan string, 1)
+	updateErrCh := make(chan error, 1)
+	newCtx, cancel := context.WithCancel(ctx)
+	var wg sync.WaitGroup
+	defer func() {
+		cancel()
+		wg.Wait()
+	}()
 
-	updatedPath, err := relaySubDirUpdated(ctx, watcherInterval, relayLogDir, fullPath, relayLogFile, latestPos)
-	if err != nil {
+	wg.Add(1)
+	go func(latestPos int64) {
+		defer func() {
+			close(switchCh)
+			close(switchErrCh)
+			wg.Done()
+		}()
+		needSwitchSubDir(newCtx, r.cfg.RelayDir, currentUUID, fullPath, latestPos, switchCh, switchErrCh)
+	}(latestPos)
+
+	wg.Add(1)
+	go func(latestPos int64) {
+		defer func() {
+			close(updatePathCh)
+			close(updateErrCh)
+			wg.Done()
+		}()
+		relaySubDirUpdated(newCtx, watcherInterval, relayLogDir, fullPath, relayLogFile, latestPos, updatePathCh, updateErrCh)
+	}(latestPos)
+
+	select {
+	case <-ctx.Done():
+		return false, false, 0, "", "", nil
+	case switchResp := <-switchCh:
```

Comment:
will relay break-then-autoresume, so generated a new subdirectory. meanwhile old subdirectory has updated something before break, so in this `select` can't assure priority?

Reply:
add a check in 20f40cd